File: Bottom View of Binary Tree.cpp
#include <vector>
#include <map>
#include <queue>
using namespace std;

struct Node {
    int data;
    Node* left;
    Node* right;
    
    Node(int key) {
        data = key;
        left = right = nullptr;
    }
};

class Solution {
public:
    vector<int> bottomView(Node* root) {
        vector<int> arr;
        if (!root) return arr;

        map<int, int> hm;
        queue<Node*> q;
        queue<int> distQ;

        q.push(root);
        distQ.push(0);

        while (!q.empty()) {
            Node* top = q.front(); q.pop();
            int dist = distQ.front(); distQ.pop();

            hm[dist] = top->data;

            if (top->left) {
                q.push(top->left);
                distQ.push(dist - 1);
            }
            if (top->right) {
                q.push(top->right);
                distQ.push(dist + 1);
            }
        }

        for (const auto& pair : hm)
            arr.push_back(pair.second);

        return arr;
    }
};

File: Bottom View of Binary Tree.java
/*
class Node {
    int data;
    Node left;
    Node right;

    Node(int data) {
        this.data = data;
        left = null;
        right = null;
    }
}
*/

class Solution {
    public class Pair{
        int dist;
        Node node;
        public Pair(Node node, int dist){
            this.dist = dist;
            this.node = node;
        }
    }
    public ArrayList<Integer> bottomView(Node root) {
        HashMap<Integer, Integer> hm = new HashMap<>();
        Queue<Pair> q = new LinkedList<>();
        q.offer(new Pair(root, 0));
        
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                Pair top = q.poll();
                hm.put(top.dist, top.node.data);
            
                Node node = top.node;
                if(node.left != null) q.offer(new Pair(node.left, top.dist - 1));
                if(node.right != null) q.offer(new Pair(node.right, top.dist + 1));
            }
        }
        
        ArrayList<Integer> ans = new ArrayList<>(hm.keySet()); 
        Collections.sort(ans);
        for(int i = 0; i < ans.size(); i++) ans.set(i, hm.get(ans.get(i)));
        
        return ans;
    }
}